Fix transport=websockets on windows.

element-82 opened the pull request:
Resolves #268
MacOS and Linux both work well with transport=websockets.  However,
windows has a different low level socket implementation.  The
client.py fails to handle zero data packets and EAGAIN correctly in
some instances.  This commit addresses the windows issues that were
seen.

Signed-off-by: Paul Barrette <paulbarrette@gmail.com>

PierreF commented:
Could you provide a documentation state that Windows socket may return 0 data packet ? That sound like a bug to me.
Python documentation explicitly state that "When a recv returns 0 bytes, it means the other side has closed" (https://docs.python.org/3/howto/sockets.html#using-a-socket).

Moreover you are not testing for a 0 [length] data packet, but None.

All of this make me think the bug it not there, by in WebsocketWrapper._recv_impl. AFAIK socket.recv *never* return None, therefor WebsocketWrapper._recv_impl should not return None.

PierreF commented:
By looking at WebsocketWrapper._recv_impl, the only two case I can see it could return None are:
* We received a Websocket frame with a payload size of 0. While this seems value for Websocket protocol itself, it seems strange for MQTT over Websocket.
* We requested a recv with a length of 0, which is not possible since the recv() is called from a "while length > 0

But not of those case are OS related, they are either a bug on paho (call recv(0)) or your broker/paho (a Websocket frame with 0 length payload look a broker bug... on the other hand paho may handle them correctly).

Could you details how to reproduce this issue ? What broker are you using ? What kind of message ? Using TLS or plain Websocket ?
If possible, could you log on WebsocketWrapper._recv_impl if it's called with length == 0 and/or if it receive an OPCODE_BINARY with payload_lenth == 0 ?

element-82 commented:
Details:

- python 3.6 on windows 10
- using TLS over websocket
- message types are json strings

I will gather as much debug logging as I can and get back to you.  BTW I can set you up with an eval account so that you can see the issue.  I don't want to send the account information here, would prefer a private email/slack or iirc.  I suspect this is a POSIX vs windows behaviour.

element-82 commented:
Datapoint:
I added a check for data == b'' before the data == None.  The new check does not hit.  Only data == None.

element-82 commented:
More data:

In _recv_impl:
```
if opcode == WebsocketWrapper.OPCODE_BINARY:
                print("OPCODE_BINARY: result = %s payload len = %d" %(result, payload_length))
                return result
```

In _packet_read
```
        while self._in_packet['to_process'] > 0:
            try:
                data = self._sock.recv(self._in_packet['to_process'])
                if data == None:
                    print ("received data == None")
```
log output:
```
OPCODE_BINARY: result = bytearray(b'0') payloadlen=136
OPCODE_BINARY: result = bytearray(b'\x85') payloadlen=136
OPCODE_BINARY: result = bytearray(b'\x01') payloadlen=136
received data == None
```

PierreF commented:
On your last comment, did you provided all log "OPCODE_BINARY" that happened before the "received data == None" ? If yes, that means recv_impl does not return None in this branch since that branch returned bytearray(b'\x01') on its last run. And its also means that bug is elsewhere in recv_impl... but I don't see other place where recv_impl could return None :/
Which version of paho-mqtt are you using ?

element-82 commented:
Yes.  I printed out all OPCODE_BINARY sections.  Here is another data point.  My theory is there is a valid pkt to read and bytes to read are not 0.  I added a print statement in the recv handler for the bytesize and when run, bytesize is > 0, but the data is still None.  That sounds like a timing issue where the data is not in cache yet.   I added a second read if the first read result is None.  E.g.:

```
def _sock_recv(self, bufsize):
    d = self._sock.recv(bufsize)
    if d == None:
        d = self._sock.recv(bufsize)
```

With this change, I do not see data == None in the _recv_impl anymore.

element-82 commented:
Cleaned up the patch to the smallest required changes.

PierreF commented:
I'm still unconvinced by the way the fix is done. In my opinion socker-like recv() method should never return None. They must always return a byte string or raise an exception. Based on Python howto socket this is also true on Windows.
Therefor for my this bug is in WebsocketWrapper, not in the loop_read method.

PierreF commented:
I've pushed a commit on fixes branch that fix the issue I've seen with empty websocket frame. Could you try to see if that fix your issue ?

element-82 commented:
Pierre,
I tried your fixes branch.  The first problem is that EAGAIN is not handled
here at client.py:
2924

>         except socket.error as err:
>             if err.errno == errno.ECONNABORTED:
>                 self.connected = False
>                 return b''
>             else:  <-----need to handle EAGAIN
>                 # no more data
>                 raise


If I add the EAGAIN exception filter, then I get the TypeError NoneType
exception again at 1842

            else:
>                 if len(data) == 0:
>                     return 1
>                 self._in_packet['to_process'] -= len(data)
>                 self._in_packet['packet'] += data


If I fix both issues on top of your fix, it works.  I think my patch was
simpler.  I think it doesn't impact anything since data should not be None,
and if it is it is handled.

Pb


On Sun, Feb 11, 2018 at 7:33 AM, Pierre Fersing <notifications@github.com>
wrote:

> I've pushed a commit on fixes branch that fix the issue I've seen with
> empty websocket frame. Could you try to see if that fix your issue ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/paho.mqtt.python/pull/269#issuecomment-364748272>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALjn5KlW-spGy6dy3o9UTPWthIKIHQcqks5tTt4JgaJpZM4Ry994>
> .
>



-- 
--
Paul Barrette
Linked In Profile <http://ca.linkedin.com/pub/paul-barrette/11/b42/a96>

Ő
 \\=^%
 (*)---(*)
^^~~^~~~^^^~~^

PierreF commented:
I don't see why EAGAIN should be handled here. It should be handled in _packet_read as for any other kind of socket (plain tcp or ssl).
I don't known how we could handle the EAGAIN in _recv_impl.

What error do you got when you run the fixes branch ?

element-82 commented:
I get the [Error 11] which is EAGAIN.

Pb

On Thu, Feb 15, 2018 at 2:40 PM, Pierre Fersing <notifications@github.com>
wrote:

> I don't see why EAGAIN should be handled here. It should be handled in
> _packet_read as for any other kind of socket (plain tcp or ssl).
> I don't known how we could handle the EAGAIN in _recv_impl.
>
> What error do you got when you run the fixes branch ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/paho.mqtt.python/pull/269#issuecomment-366038943>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALjn5CYvVNnWXwbimIT5v5zJsc9aecorks5tVIhLgaJpZM4Ry994>
> .
>



-- 
--
Paul Barrette
Linked In Profile <http://ca.linkedin.com/pub/paul-barrette/11/b42/a96>

Ő
 \\=^%
 (*)---(*)
^^~~^~~~^^^~~^

element-82 commented:
Hi Pierre,
any more thoughts on this?  Anything you would like me to try?
Pb

On Fri, Feb 16, 2018 at 1:29 PM, Paul Barrette <paulbarrette@gmail.com>
wrote:

> I get the [Error 11] which is EAGAIN.
>
> Pb
>
> On Thu, Feb 15, 2018 at 2:40 PM, Pierre Fersing <notifications@github.com>
> wrote:
>
>> I don't see why EAGAIN should be handled here. It should be handled in
>> _packet_read as for any other kind of socket (plain tcp or ssl).
>> I don't known how we could handle the EAGAIN in _recv_impl.
>>
>> What error do you got when you run the fixes branch ?
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/eclipse/paho.mqtt.python/pull/269#issuecomment-366038943>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ALjn5CYvVNnWXwbimIT5v5zJsc9aecorks5tVIhLgaJpZM4Ry994>
>> .
>>
>
>
>
> --
> --
> Paul Barrette
> Linked In Profile <http://ca.linkedin.com/pub/paul-barrette/11/b42/a96>
>
> Ő
>  \\=^%
>  (*)---(*)
> ^^~~^~~~^^^~~^
>
>


-- 
--
Paul Barrette
Linked In Profile <http://ca.linkedin.com/pub/paul-barrette/11/b42/a96>

Ő
 \\=^%
 (*)---(*)
^^~~^~~~^^^~~^

PierreF commented:
Thank for your patience. I've seen that my previous fix couldn't work on Windows (EAGAIN is not errno.EAGAIN in Windows).
Pushed another commit that hopefully will fix this issue.

element-82 commented:
Pierre,
that fixed the issue.
Thanks
Pb

PierreF commented:
Closing as the issue is now fixed
